### PR TITLE
wallet: Releases dangling files on BerkeleyEnvironment::Close

### LINKED ISSUE #15291: Bitcoin-core multiwallets limited to about 300 loads
Hi,
Using Bitcoin Core RPC client version v0.17.0.1-g1476554d3b1db5352c0de1de7524864d4bce3aac

I'm using bitcoin-cli loadwallet and unloadwallet commands. They both seem to work fine.
Unfortunately, memory is getting higher and higher, and no wallet can be loaded after 300 have been.
The problem may be related to BerkeleyDB rather than to Bitcoin Core itself. Fact is that lsof shows handles not cleared off after the unloadwallet command is issued. Files still handled are .walletlock, db.log and database/log.0000000001 .
When 300 different wallets have been loaded/unloaded, used RAM is over 8 GB, and bitcoind will finally become unstable and fail.

Any help would be appreciated.
Regards,

### PR DESCRIPTION
**Pull request by promag**
This PR closes `db.log` and removes `.walletlock` files when `BerkeleyEnvironment` is closed.

Fixes https://github.com/bitcoin/bitcoin/issues/15291#issuecomment-459131886.

### CODE REVIEW

**Comment by gmaxwell** [comment]
Concept ACK

**Comment by promag** [comment]
Please add to 0.18 milestone.

File: src/wallet/db.cpp [hunk 1]
```diff
@@ -126,6 +126,10 @@ void BerkeleyEnvironment::Close()
         }
     }
 
+    FILE* error_file = nullptr;
+    dbenv->get_errfile(&error_file);
+    if (error_file) fclose(error_file);
```
Comment (promag) [inline src/wallet/db.cpp]: Maybe this should be after `dbenv->close(0)`.

**Comment by MarcoFalke** [comment]
Needs rebase

**Comment by promag** [comment]
Rebased.

**Comment by practicalswift** [comment]
Concept ACK

**Review by ryanofsky — APPROVED** [review]
utACK 2f8b8f479bb43729ca2ff40929e8463347b0b7b4. Nice cleanup!

**Comment by promag** [comment]
Thanks @ryanofsky, pushed a test that fails without this PR.

**Review by ryanofsky — APPROVED** [review]
utACK d3bf3b930d34da7d121ae35b4fb75865ed73208c. Just added test since the previous review.

**Comment by laanwj** [comment]
utACK 2f8b8f479bb43729ca2ff40929e8463347b0b7b4

**Comment by fanquake** [comment]
utACK d3bf3b9

master (fc21bb4):
```
src/qt/bitcoin-qt

lsof -p 56700 | grep wallet
bitcoin-q 56700 michael   24u     REG                1,4        0  4556196 /Users/michael/Library/Application Support/Bitcoin/wallets/.walletlock
bitcoin-q 56700 michael   25w     REG                1,4        0  4556198 /Users/michael/Library/Application Support/Bitcoin/wallets/db.log
bitcoin-q 56700 michael   33u     REG                1,4  1048576 26847564 /Users/michael/Library/Application Support/Bitcoin/wallets/database/log.0000000001
bitcoin-q 56700 michael   39w     REG                1,4        0  4556198 /Users/michael/Library/Application Support/Bitcoin/wallets/db.log

src/bitcoin-cli unloadwallet ""

lsof -p 56700 | grep wallet
bitcoin-q 56700 michael   24u     REG                1,4        0  4556196 /Users/michael/Library/Application Support/Bitcoin/wallets/.walletlock
bitcoin-q 56700 michael   25w     REG                1,4        0  4556198 /Users/michael/Library/Application Support/Bitcoin/wallets/db.log
bitcoin-q 56700 michael   39w     REG                1,4        0  4556198 /Users/michael/Library/Application Support/Bitcoin/wallets/db.log
```

This PR:
```
src/qt/bitcoin-qt

lsof -p 54793 | grep wallet
bitcoin-q 54793 michael   36u      REG                1,4        0  4556196 /Users/michael/Library/Application Support/Bitcoin/wallets/.walletlock
bitcoin-q 54793 michael   37w      REG                1,4        0  4556198 /Users/michael/Library/Application Support/Bitcoin/wallets/db.log
bitcoin-q 54793 michael   38u      REG                1,4  1048576 26847564 /Users/michael/Library/Application Support/Bitcoin/wallets/database/log.0000000001

src/bitcoin-cli unloadwallet ""

lsof -p 54793 | grep wallet
```

**Comment by MeshCollider** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/15297/commits/d3bf3b930d34da7d121ae35b4fb75865ed73208c

**Comment by promag** [comment]
Should be backport?

**Comment by fanquake** [comment]
Being backported in #15575.

**Pull request merged**